azkaban-common/src/main/java/azkaban/executor/ExecutionController.java

```diff
@@ -407,14 +410,33 @@ public String getJobLinkUrl(final ExecutableFlow exFlow, final String jobId, fin
   }
 
   /**
-   * If a flow was dispatched to an executor, cancel by calling Executor. Else if it's still
-   * queued in database, remove it from database queue and finalize. {@inheritDoc}
+   * If a flow is already dispatched to an executor, cancel by calling Executor. Else if it's still
+   * queued in DB, remove it from DB queue and finalize. {@inheritDoc}
    */
   @Override
   public void cancelFlow(final ExecutableFlow exFlow, final String userId)
       throws ExecutorManagerException {
-    // Todo: call executor to cancel the flow if it's running or remove from DB queue if it
-    // hasn't started
+    synchronized (exFlow) {
```

Comment:
Discussed offline. The entire canceling is synchronized to prevent two different threads finalizing the flow in DB as well as calling executor to cancel at the same time. To be safe, the second thread should wait for the first thread to finish canceling first.